### PR TITLE
[Generation] Fix Transition probs

### LINKED ISSUE #15869: Mismatch between beam search score transition probabilities and beam sequence scores
## Environment info

- `transformers` version: 4.16.2
- Platform: Linux-5.13.0-30-generic-x86_64-with-glibc2.31
- Python version: 3.9.7
- PyTorch version (GPU?): 1.10.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help

I'm tagging @patrickvonplaten because he has recently worked on this #14654

## Information

The model is a __T5__. I'm using its conditional generation properties and testing beam search decoding outputs:

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

Purpose: Create two simple dummy events and test whether the transition probability scores are the same as the probability sequence scores. The goal is to understand what the sequence scores represent (are they unnormalized by length? normalized by length?). From the PR #14654 discussion, I had the impression that it would be enough to sum the transition probabilities but these do not seem to match. Would you please help me understand?

## To reproduce

1. Create two simple dummy input examples. 
2. Encode them using a t5 model (`model`) and generate the output scores using beam search.
3. Obtain transition scores through the use of `model.compute_transition_beam_scores`
4. Sum the results of 3 and compare with the ones returned with the `BeamSearchEncoderDecoderOutput.sequence_scores`.

```python
import transformers

model_name = "t5-small"
encoding_hyperparameters = {
    "padding": "max_length",
    "max_length": 512,
    "truncation": True,
    "add_special_tokens": True,
    "return_attention_mask": True,
    "return_tensors": "pt",
}

tokenizer = transformers.T5TokenizerFast.from_pretrained(model_name)
model = transformers.T5ForConditionalGeneration.from_pretrained(model_name)

EXAMPLE = ["question: How are you? \n context: I had a long day, she said. I am so exhausted.", 
           "question: What did the fox do? \n context: The fox jumped over the fence into a very green lawn."]

BEAM_SEARCH_KWARGS = {
    "num_beams": 4,
    "do_sample": False,
    "num_return_sequences": 1,
}

# Encode inputs
inputs_ids = tokenizer(EXAMPLE, **encoding_hyperparameters)

# Generate using beam search
beamsearch_results = model.generate(
    input_ids=inputs_ids["input_ids"],
    attention_mask=inputs_ids["attention_mask"],    
    max_length=10,
    return_dict_in_generate=True,
    output_scores=True,
    # the id of the token to force as the last generated token when max_length is reached
    forced_eos_token_id=tokenizer.eos_token_id,
    **BEAM_SEARCH_KWARGS
)

trs_bs = model.compute_transition_beam_scores(
    sequences=beamsearch_results.sequences,
    scores=beamsearch_results.scores, 
    beam_indices=beamsearch_results.beam_indices
)

print("Summ:", torch.sum(trs_bs, dim=1), "Expected:", beamsearch_results.sequences_scores)
print("Sum/length:", torch.sum(trs_bs, dim=1)/beamsearch_results.beam_indices.shape[-1], "Expected:", beamsearch_results.sequences_scores)
# output
# Sum: tensor([-1.5411, -0.3851]) Expected: tensor([-0.1712, -0.0428])
# Sum/length: tensor([-0.1712, -0.0428]) Expected: tensor([-0.1712, -0.0428])

```

From the example above I deduced that in order to obtain the same scores as those computed in the `sequences_scores` it would suffice to divide by the length of the sentences. In this case, it seems to work nicely because both sequences have the same length:
```python
# output of beamsearch_results.sequences
tensor([[    0,    27,   141,     3,     9,   307,   239,     6,   255,     1],
        [    0,     3, 16287,   147,     8,  8227,   139,     3,     9,     1]])
```

So I tried a different example, that would cause the beamsearch_results.sequences to be different: 
```python
# Example 2
# The only change to the script above is the example, where we modify the first sequence in the batch
EXAMPLE = ["question: Is this yes or no question? \n context: It is yes", 
           "question: What did the fox do? \n context: The fox jumped over the fence into a very green lawn."]
# ...

print("Sum:", torch.sum(trs_bs, dim=1), "Expected:", beamsearch_results.sequences_scores)
print("Sum/length:", torch.sum(trs_bs, dim=1)/beamsearch_results.beam_indices.shape[-1], "Expected:", beamsearch_results.sequences_scores)
print("Sum/rel_length:", torch.sum(trs_bs, dim=1) / torch.sum(trs_bs != 0, dim=1), "Expected:", beamsearch_results.sequences_scores)
# outputs
# Sum: tensor([-0.0770, -0.3851]) Expected: tensor([-0.0385, -0.0428])
# Sum/length: tensor([-0.0086, -0.0428]) Expected: tensor([-0.0385, -0.0428])
# Sum/rel_length: tensor([-0.0385, -0.0481]) Expected: tensor([-0.0385, -0.0428])
```

The output of `beamsearch_results.sequences` for the above example is: 
```python
tensor([[    0,  4273,     1,     0,     0,     0,     0,     0,     0,     0],
        [    0,     3, 16287,   147,     8,  8227,   139,     3,     9,     1]])
```

The difference from `Sum/length` to `Sum/rel_length` is that in the former I divide by the maximum length of the generated sentences, whereas the previous is divided by the number of non-zero transition probabilities. We can see that for the latter case, (i.e., when dividing by the relative length) only the first example score is matched to the original `beamsearch_results.sequences_scores`).

Will you please help me better understand the computation of these probabilities and their connection with the sequence_scores? In particular, are the individual scores returned by the `compute_transition_beam_scores` length-normalized ?  Do these individual scores aim to represent the joint probability or are they representing the individual probabilities? Are we supposed to consider the initial padding token when computing the scores? 

Thanks in advance for your time!

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes #15869

This PR fixes incorrectly computed transition probabilties for beam search.
In beam search it's common to want to know the transition probability between tokens -> see: https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175/15?u=patrickvonplaten

Currently there is a bug for every beam search that ends before `max_length` as discovered by #15869 .

This PR fixes this behavior and adds a couple of tests. 

🚨 **There is a tiny breaking change** in the output format of `beam_indices`. Instead of returning a `tuple(tuple(int))` a `torch.LongTensor` is returned. Since the feature was broken before (`beam_indices` were incorrect), this is hardly a breaking change and necessary to make the functionality more userfriendly and robustu🚨

@patil-suraj @gante , it would be super nice if you could do an in-depth review here also to understand this feature.
**Note:** people do seem to be interested in output scores for generation as can be seen by the large number of views here: https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175

Feel free to drop any question in the PR if you don't understand something or something is unclear - thanks :pray:

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
I'll talk about this feature in more-detail in a tutorial

**Review by sgugger — APPROVED** [review]
Thanks for adding this with new tests! LGTM at a high level :-)

**Review by gante — APPROVED** [review]
LGTM! 👍

(I see from this PR that TF is lacking many useful things like `sequence_scores`, which are not being set)

**Comment by adamkhakhar** [comment]
Hello,
Regarding the follow up to these 3 posts: https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175/23, https://discuss.huggingface.co/t/retrieving-probability-over-tokens-during-beam-search/21217, https://discuss.huggingface.co/t/get-top-k-tokens-for-each-time-step-instead-of-the-highest-probability-token/14032:

Can you show us, step by step, how to retrieve the probability for each token generated for each prediction in a `transformers.generation_utils.BeamSearchEncoderDecoderOutput`?

Much appreciated.

**Comment by gante** [comment]
Hi @adamkhakhar -- it is in our plans to write a tutorial for that :) It probably won't happen in the next 1 or 2 months, but stay tuned 🙏

**Comment by MotzWanted** [comment]
Hi @patrickvonplaten and @gante; what is the perspective on the tutorial?

**Comment by gante** [comment]
No news yet :)

**Comment by MotzWanted** [comment]
Thanks for the quick response @gante. I will stay tuned. Can you tell me whether you can expect the first generated sequence to always have the highest probability?

**Comment by gante** [comment]
@MotzWanted yeah, if you use `num_return_sequences >1` with beam search, the output sequences are sorted by their score (from highest to lowest)

**Comment by MotzWanted** [comment]
But if I replicate this [showcase](https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175/26) of @patrickvonplaten to compute the sequence probabilities of the output scores, I don't get that output sequences are sorted by their probability (from highest to lowest). Can you explain to me why this is so @gante?

**Comment by gante** [comment]
@MotzWanted my reply was given in the context of this thread, which is related to `beam_search`. Any call with `do_sample=True`, as in the example you linked, will lose the sorting property.
